text/0000-float-semantics.md

```diff
@@ -0,0 +1,287 @@
+
+The non-deterministic choice happens when the operation is executed; i.e., the result of a NaN-producing floating point operation is a stable bit pattern (looking at these bits multiple times will yield consistent results), but running the same operation twice with the same inputs can produce different results.
+
+Unless noted otherwise, the same rules also apply to NaNs returned by other library functions (e.g. `min`, `minimum`, `max`, `maximum`); other aspects of their semantics and which IEEE 754-2008 operation they correspond to are documented with the respective functions.
+
+### `const` semantics
+
+Const-evaluation must necessarily be deterministic to ensure soundness of the type system.
+`const` use of floating points does not make any guarantees beyond that:
+when a floating-point operation produces a NaN result, the resulting NaN bit pattern is *some* deterministic function of the operation's inputs that satisfies the constraints placed on run-time floating point semantics.
+However, the exact function is not specified, and it is allowed to change across targets and Rust versions, and even with compiler flags.
+In particular, there is no guarantee that the choice made in const evaluation is consistent with the choice made at runtime.
+That is, the following assertion is allowed to fail (and in fact, it [fails on current versions of Rust](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a594d2975c29b1c7fa457a4ec4ae4b87)):
+```rust
+#![allow(non_snake_case)]
+use std::hint::black_box;
+
+const C: f32 = 0.0 / 0.0;
+
+fn main() {
+    let c: f32 = 0.0 / black_box(0.0);
+    assert_eq!(C.to_bits(), c.to_bits());
+}
+```
+
+This means that evaluating the same `const fn` on the same arguments can produce different results at compile-time and run-time.
+However, note that these functions are already non-deterministic: even evaluating the same function with the same arguments twice at runtime can [and does](https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=50b5a549fa1fe259cea5ad138066ccf0) produce different results!
+
+This resolves the last open question blocking floating-point operations in `const fn`.
+When the RFC is accepted, the `const_fn_floating_point_arithmetic` feature gate and the `const fn to_bits` methods can be stabilized.
+
+### Assumptions about floating-point environment
+
+This RFC is primarily concerned with the guarantee Rust provides to its users.
+How exactly those guarantees are achieved is an implementation detail, and not observable when writing pure Rust code.
+However, when mixing Rust with inline assembly, those details *do* become observable.
+To ensure that Rust can provide the above guarantees to user code, it is UB for inline assembly to alter the behavior of floating-point operations in any way: when leaving the inline assembly block, the floating-point environment must be in exactly the same state as when the inline assembly block was entered.
```

Comment:
I think this should be clarified to state that it's fine to change the floating-point environment inside the inline asm, but that it needs to be restored before reentering any Rust/C/etc. code. Also, we'll probably want to state that this is referring specifically to fp control flags, changing the fp status bits is fine (maybe only if they're properly marked as clobbered or something).

Reply:
> I think this should be clarified to state that it's fine to change the floating-point environment inside the inline asm, but that it needs to be restored before reentering any Rust/C/etc. code.

It already clarifies:

"when leaving the inline assembly block, the floating-point environment must be in exactly the same state as when the inline assembly block was entered."

I don't think there is any need for further clarification here.